File: mini-claw-code-starter-ts/src/tools/edit.ts
import { readFile, writeFile } from "node:fs/promises";

import type { JsonValue, Tool } from "../types";
import { ToolDefinition } from "../types";

export class EditTool implements Tool {
  readonly toolDefinition: ToolDefinition;

  constructor() {
    this.toolDefinition = ToolDefinition.new(
      "edit",
      "Replace an exact string in a file (must appear exactly once).",
    )
      .param("path", "string", "The file path to edit", true)
      .param("oldString", "string", "The exact string to find and replace", true)
      .param("newString", "string", "The replacement string", true);
  }

  static new(): EditTool {
    return new EditTool();
  }

  definition(): ToolDefinition {
    return this.toolDefinition;
  }

  async call(_args: JsonValue): Promise<string> {
    void readFile;
    void writeFile;
    throw new Error(
      "TODO(ch4): read the file, ensure oldString appears exactly once, write the replacement, and confirm success",
    );
  }
}

File: mini-claw-code-starter-py/src/mini_claw_code_starter_py/tools/edit.py
from __future__ import annotations

from typing import Any

from ..types import ToolDefinition


class EditTool:
    def __init__(self) -> None:
        self._definition = None

    @property
    def definition(self) -> ToolDefinition:
        if self._definition is None:
            raise RuntimeError("EditTool.new() must initialize self._definition")
        return self._definition

    @classmethod
    def new(cls) -> "EditTool":
        raise NotImplementedError(
            "Define an edit tool with required 'path', 'old_string', and 'new_string' parameters"
        )

    async def call(self, args: Any) -> str:
        raise NotImplementedError(
            "Replace exactly one occurrence of old_string in the target file and return a confirmation"
        )
